File: output_20260318_123118/qwen3/c/Struct/account.c
/*@ predicate account(struct account *a; int balance) = a->balance |-> balance; @*/

struct account
{
    int balance;
};

//@ requires true;
//@ ensures account(result, 0);
struct account *create_account()
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    myAccount->balance = 0;
    return myAccount;
}

//@ requires account(myAccount, ?old_balance);
//@ ensures account(myAccount, newBalance);
void account_set_balance(struct account *myAccount, int newBalance)
{
    myAccount->balance = newBalance;
}

//@ requires account(myAccount, ?balance);
//@ ensures true;
void account_dispose(struct account *myAccount)
{
    free(myAccount);
}

//@ requires true;
//@ ensures true;
int main()
{
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_dispose(myAccount);
    return 0;
}

File: output_repair/qwen3/sample_3/repair_3/rust/rust/purely_unsafe/solutions/account.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};
use std::ptr;

struct Account {
    balance: i32,
}

predicate account_owned(ptr: *mut Account) =
    ptr != ptr::null_mut() &*&
    heap_block(ptr as *mut u8, std::mem::size_of::<Account>());

impl Account {
    #[ensures(account_owned(result))]
    unsafe fn create() -> *mut Account {
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        (*my_account).balance = 0;
        my_account
    }

    #[requires(account_owned(my_account))]
    #[ensures(account_owned(my_account))]
    unsafe fn set_balance(my_account: *mut Account, new_balance: i32) {
        (*my_account).balance = new_balance;
    }

    #[requires(account_owned(my_account))]
    unsafe fn dispose(my_account: *mut Account) {
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }
}

fn main() {
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::dispose(my_account);
    }
}